### PR TITLE
Expose `Colormap.num_colors` and document usage for cycling through qualitative colormaps 

### LINKED ISSUE #54: Feature: improve ergonomics for cyclic qualitative/categorical colormaps
* cmap version: 0.2.0
* Python version: 3.9.18
* Operating System: macOS 14.4.1

### Description

I am using cmap to generate a qualitative colormap for integer valued images in a napari plugin. The integer values can be greater than the number of colors in the qualitative/categorical colormap, so I want to cycle through those colors.

I might be missing something in the API, but I found this to be a little awkward right now. I think I need to do something like the following:

```python
from cmap import Colormap
index: int = 0 # but could take any non-negative integer value
colormap = Colormap("colorbrewer:set1_8")
color = colormap(index % len(colormap.color_stops))
```

The main reason I found this awkward is because it is was hard to find `Colormap.color_stops`. Alternatively, I also found `len(tuple(Colormap.iter_colors()))`, but this felt wasteful and even more verbose.

For more context, [see the equivalent lines at a commit in a recent PR](https://github.com/chanzuckerberg/napari-cryoet-data-portal/pull/21/files#diff-c5d5e3e005e0b5e48111e418a2702d9bcc1e0a32d02ed811b858018f08dfc515R27).

### Ideas

I'm not sure what to suggest here since I only just started using `cmap` and I don't want to pollute what is a very concise API.

Adding something like `Colormap.num_colors` is probably enough to make me happy here (where `num_colors` is equal to `len(colormap.color_stops)`).

I also thought of something like `Colormap.cycle(values)` as an alternative to `Colormap.__call__` that would do this automatically, but not sure if I like that either.

### Summary

This is a super nit-picky feature request, so you should feel free to close and move on. The only way I could justify putting more thought in here is if the cyclic mapping is one that is likely to used frequently.

If there is interest, I'm happy to think a bit more and have a go at an implementation.

Otherwise, I am loving cmap and its docsite - the [catalog](https://cmap-docs.readthedocs.io/en/stable/catalog/) and perceptual plots are beautiful!

### PR DESCRIPTION
**Pull request by andy-sweet**
Closes #54

Here's a quick attempt at the basic additions described in #54. Along with a rough attempt at document usage of qualitative colormaps more generally.

Not quite ready for review, but any feedback about anything is welcome - structure and style included.

As I wrote this, I do feel a stronger need for something like `Colormap(x, cycle=True)` or `Colormap.cycle(x)` to handle iterable/ndarray inputs, but haven't gone there yet.

### CODE REVIEW

File: docs/colormaps.md [hunk 1]
```diff
@@ -155,6 +155,105 @@ A [`numpy.ndarray`][], in one of the following formats:
 
  ... TODO
 
+### Using qualitative colormaps
```
Comment (andy-sweet) [inline docs/colormaps.md:158]: Kinda awkward this follows a TODO, but this felt like the right place for this. Other suggestions welcome!

Comment (tlambert03) [reply docs/colormaps.md:158]: haha.  a reminder to never put a "todo" in a public place :)

this is a fine place for all this.  and you could remove the TODO if you want (the blank canvas is even less helpful than nothing)

File: docs/colormaps.md [hunk 2]
```diff
@@ -155,6 +155,105 @@ A [`numpy.ndarray`][], in one of the following formats:
...
+```python
+c(9 % c.num_colors)
+```
+
+which now maps to the first color
+
+
+```python
+Color((0.8941, 0.102, 0.1098))
+```
```
Comment (tlambert03) [inline docs/colormaps.md:255]: This is an elegant and simple enough approach that I'm not sure we even need a convenience method?  You could mention that It vectorizes nicely as well:

```python
In [12]: c(np.arange(16) % len(c.color_stops))
Out[12]:
array([[0.89411765, 0.10196078, 0.10980392, 1.        ],
       [0.21568627, 0.49411765, 0.72156863, 1.        ],
       [0.30196078, 0.68627451, 0.29019608, 1.        ],
       [0.59607843, 0.30588235, 0.63921569, 1.        ],
       [1.        , 0.49803922, 0.        , 1.        ],
       [1.        , 1.        , 0.2       , 1.        ],
       [0.65098039, 0.3372549 , 0.15686275, 1.        ],
       [0.96862745, 0.50588235, 0.74901961, 1.        ],
       [0.89411765, 0.10196078, 0.10980392, 1.        ],
       [0.21568627, 0.49411765, 0.72156863, 1.        ],
       [0.30196078, 0.68627451, 0.29019608, 1.        ],
       [0.59607843, 0.30588235, 0.63921569, 1.        ],
       [1.        , 0.49803922, 0.        , 1.        ],
       [1.        , 1.        , 0.2       , 1.        ],
       [0.65098039, 0.3372549 , 0.15686275, 1.        ],
       [0.96862745, 0.50588235, 0.74901961, 1.        ]])
```

And, a side note (and perhaps a "gotcha") worth remembering here.  The dtype of the value passed to `__call__` does matter (just like with matplotlib).  If you pass in a float array, it expects the values to be betwee 0-1.  So the same in float space would be:

```python
In [31]: c(np.linspace(0, 1.999, 16) % 1)
Out[31]:
array([[0.89411765, 0.10196078, 0.10980392, 1.        ],
       [0.21568627, 0.49411765, 0.72156863, 1.        ],
       [0.30196078, 0.68627451, 0.29019608, 1.        ],
       [0.59607843, 0.30588235, 0.63921569, 1.        ],
       [1.        , 0.49803922, 0.        , 1.        ],
       [1.        , 1.        , 0.2       , 1.        ],
       [0.65098039, 0.3372549 , 0.15686275, 1.        ],
       [0.96862745, 0.50588235, 0.74901961, 1.        ],
       [0.89411765, 0.10196078, 0.10980392, 1.        ],
       [0.21568627, 0.49411765, 0.72156863, 1.        ],
       [0.30196078, 0.68627451, 0.29019608, 1.        ],
       [0.59607843, 0.30588235, 0.63921569, 1.        ],
       [1.        , 0.49803922, 0.        , 1.        ],
       [1.        , 1.        , 0.2       , 1.        ],
       [0.65098039, 0.3372549 , 0.15686275, 1.        ],
       [0.96862745, 0.50588235, 0.74901961, 1.        ]])
```

Comment (andy-sweet) [reply docs/colormaps.md:255]: > This is an elegant and simple enough approach that I'm not sure we even need a convenience method? You could mention that It vectorizes nicely as well:

Sorry, I'm a derp and temporarily forgot `%` would work fine with an `ndarray`. It doesn't work with the iterable, but at that point we could just tell users to coerce to an `ndarray` at this time.

I'll add some docs like the example above.

> And, a side note (and perhaps a "gotcha") worth remembering here. The dtype of the value passed to __call__ does matter (just like with matplotlib). If you pass in a float array, it expects the values to be betwee 0-1. So the same in float space would be:

Good point. I'll reiterate that here.

**Comment by andy-sweet** [comment]
Addressed the PR comments. The codecov patch check failed, but given the code changes here, I don't see a need to add more tests.

File: docs/colormaps.md [hunk 3]
```diff
@@ -151,9 +151,143 @@ A [`numpy.ndarray`][], in one of the following formats:
...
+       [0.65098039, 0.3372549 , 0.15686275, 1.        ],
+       [0.96862745, 0.50588235, 0.74901961, 1.        ]])
+```
+
+The behavior of calling a `Colormap` with an array depends on its `dtype`.
+With a floating point `dtype`, it expects the values to be [0, 1], so the
+equivalent call to the above is
+
+```python
+c(np.linspace(0, 2, 16, endpoint=False) % 1)
```
Comment (tlambert03) [inline docs/colormaps.md:286]: `endpoint=False` 🤯  :joy:  can't believe i hadn't found that

**Comment by tlambert03** [comment]
thanks @andy-sweet!

**Pull request merged**